water_neuron/src/main.rs

```diff
@@ -333,6 +335,30 @@ async fn icp_to_nicp(arg: ConversionArg) -> Result<DepositSuccess, ConversionErr
     check_postcondition(water_neuron::conversion::icp_to_nicp(arg).await)
 }
 
+#[update]
+async fn cancel_withdrawal(neuron_id: NeuronId) -> Result<MergeResponse, String> {
+    reject_anonymous_call();
+
+    let neuron_ids: Vec<Option<NeuronId>> = read_state(|s| {
+        s.account_to_withdrawals
+            .get(&ic_cdk::caller().into())
+            .cloned()
+            .unwrap_or(vec![])
+            .iter()
+            .map(|id| s.withdrawal_id_to_request.get(id).unwrap().neuron_id)
+            .collect()
+    });
+
+    for maybe_neuron_id in neuron_ids {
+        if let Some(target_neuron_id) = maybe_neuron_id {
+            if target_neuron_id == neuron_id {
+                return water_neuron::conversion::cancel_withdrawal(neuron_id).await;
```

Comment:
what's the point of this loop?

Reply:
I made it to check that the call is the owner of the neuron. Otherwise anybody could cancel the withdrawal

Reply:
in the state, if you look at `neuron_id_to_withdrawal_id` and `withdrawal_id_to_request` and `principal_to_withdrawal` you can check the owner of the neuron and verify that he's allowed to cancel.